src/TurbulenceClosures/turbulence_closure_implementations/lagrangian_dynamic_smagorinsky.jl

```diff
@@ -0,0 +1,319 @@
+
+    # Averaging over a 27-point stencil
+    # Remember! This is not a box-filter, more
+    # of a gaussian filter
+    # TODO: explore different filters
+    u̅  = ℑxᶜᵃᵃ(i, j, k, grid, ℑxyzᶠᶜᶜ, ℑxyzᶜᶠᶠ, ℑxyzᶠᶜᶜ, ℑxyzᶜᶠᶠ, u)
+    v̅  = ℑyᵃᶜᵃ(i, j, k, grid, ℑxyzᶜᶠᶜ, ℑxyzᶠᶜᶠ, ℑxyzᶜᶠᶜ, ℑxyzᶠᶜᶠ, v)
+    w̅  = ℑzᵃᵃᶜ(i, j, k, grid, ℑxyzᶜᶜᶠ, ℑxyzᶠᶠᶜ, ℑxyzᶜᶜᶠ, ℑxyzᶠᶠᶜ, w)
+
+    u̅u̅ = ℑxᶜᵃᵃ(i, j, k, grid, ℑxyzᶠᶜᶜ, ℑxyzᶜᶠᶠ, ℑxyzᶠᶜᶜ, ℑxyzᶜᶠᶠ, uᵢuⱼ, u, u)
+    v̅v̅ = ℑyᵃᶜᵃ(i, j, k, grid, ℑxyzᶜᶠᶜ, ℑxyzᶠᶜᶠ, ℑxyzᶜᶠᶜ, ℑxyzᶠᶜᶠ, uᵢuⱼ, v, v)
+    w̅w̅ = ℑzᵃᵃᶜ(i, j, k, grid, ℑxyzᶜᶜᶠ, ℑxyzᶠᶠᶜ, ℑxyzᶜᶜᶠ, ℑxyzᶠᶠᶜ, uᵢuⱼ, w, w)
+
+    u̅v̅ = ℑxyzᶜᶜᶜ(i, j, k, grid, ℑxyzᶠᶠᶠ, ℑxyzᶜᶜᶜ, ℑxyzᶠᶠᶠ, uᵢuⱼ, ℑxᶜᵃᵃ, ℑyᵃᶜᵃ, u, v)
+    u̅w̅ = ℑxyzᶜᶜᶜ(i, j, k, grid, ℑxyzᶠᶠᶠ, ℑxyzᶜᶜᶜ, ℑxyzᶠᶠᶠ, uᵢuⱼ, ℑxᶜᵃᵃ, ℑzᵃᵃᶜ, u, w)
+    v̅w̅ = ℑxyzᶜᶜᶜ(i, j, k, grid, ℑxyzᶠᶠᶠ, ℑxyzᶜᶜᶜ, ℑxyzᶠᶠᶠ, uᵢuⱼ, ℑyᵃᶜᵃ, ℑzᵃᵃᶜ, v, w)
+
+    S̅ₘ = ℑxyzᶜᶜᶜ(i, j, k, grid, ℑxyzᶠᶠᶠ, ℑxyzᶜᶜᶜ, ℑxyzᶠᶠᶠ, ϕ⁰⁵, ΣᵢⱼΣᵢⱼᶜᶜᶜ, u, v, w)
+
+    S̅₁₁ = ℑxyzᶜᶜᶜ(i, j, k, grid, ℑxyzᶠᶠᶠ, ℑxyzᶜᶜᶜ, ℑxyzᶠᶠᶠ, Σ₁₁, u) # Directly at centers
+    S̅₂₂ = ℑxyzᶜᶜᶜ(i, j, k, grid, ℑxyzᶠᶠᶠ, ℑxyzᶜᶜᶜ, ℑxyzᶠᶠᶠ, Σ₂₂, v) # Directly at centers
+    S̅₃₃ = ℑxyzᶜᶜᶜ(i, j, k, grid, ℑxyzᶠᶠᶠ, ℑxyzᶜᶜᶜ, ℑxyzᶠᶠᶠ, Σ₃₃, w) # Directly at centers
+    
+    S̅₁₂ = ℑxyᶜᶜᵃ(i, j, k, grid, ℑxyzᶠᶠᶜ, ℑxyzᶜᶜᶠ, ℑxyzᶠᶠᶜ, ℑxyzᶜᶜᶠ, Σ₁₂, u, v) # originally at ffc
+    S̅₁₃ = ℑxzᶜᵃᶜ(i, j, k, grid, ℑxyzᶠᶜᶠ, ℑxyzᶜᶠᶜ, ℑxyzᶠᶜᶠ, ℑxyzᶜᶠᶜ, Σ₁₃, u, w) # originally at fcf
+    S̅₂₃ = ℑyzᵃᶜᶜ(i, j, k, grid, ℑxyzᶜᶠᶠ, ℑxyzᶠᶜᶜ, ℑxyzᶜᶠᶠ, ℑxyzᶠᶜᶜ, Σ₁₂, u, v) # originally at cff
+
+    S̅S̅₁₁ = ℑxyzᶜᶜᶜ(i, j, k, grid, ℑxyzᶠᶠᶠ, ℑxyzᶜᶜᶜ, ℑxyzᶠᶠᶠ, SₘSᵢⱼ, Σ₁₁, u, ϕ⁰⁵, ΣᵢⱼΣᵢⱼᶜᶜᶜ, u, v, w)
+    S̅S̅₂₂ = ℑxyzᶜᶜᶜ(i, j, k, grid, ℑxyzᶠᶠᶠ, ℑxyzᶜᶜᶜ, ℑxyzᶠᶠᶠ, SₘSᵢⱼ, Σ₂₂, v, ϕ⁰⁵, ΣᵢⱼΣᵢⱼᶜᶜᶜ, u, v, w)
+    S̅S̅₃₃ = ℑxyzᶜᶜᶜ(i, j, k, grid, ℑxyzᶠᶠᶠ, ℑxyzᶜᶜᶜ, ℑxyzᶠᶠᶠ, SₘSᵢⱼ, Σ₃₃, w, ϕ⁰⁵, ΣᵢⱼΣᵢⱼᶜᶜᶜ, u, v, w)
+    
+    S̅S̅₁₂ = ℑxyzᶜᶜᶜ(i, j, k, grid, ℑxyzᶠᶠᶠ, ℑxyzᶜᶜᶜ, ℑxyzᶠᶠᶠ, SₘSᵢⱼ, ℑΣ₁₂, u, v, ϕ⁰⁵, ΣᵢⱼΣᵢⱼᶜᶜᶜ, u, v, w)
+    S̅S̅₁₃ = ℑxyzᶜᶜᶜ(i, j, k, grid, ℑxyzᶠᶠᶠ, ℑxyzᶜᶜᶜ, ℑxyzᶠᶠᶠ, SₘSᵢⱼ, ℑΣ₁₃, u, w, ϕ⁰⁵, ΣᵢⱼΣᵢⱼᶜᶜᶜ, u, v, w)
+    S̅S̅₂₃ = ℑxyzᶜᶜᶜ(i, j, k, grid, ℑxyzᶠᶠᶠ, ℑxyzᶜᶜᶜ, ℑxyzᶠᶠᶠ, SₘSᵢⱼ, ℑΣ₂₃, v, w, ϕ⁰⁵, ΣᵢⱼΣᵢⱼᶜᶜᶜ, u, v, w)
+
+    Δᶠ  = filter_width(i, j, k, grid)
+    TΔ  = 3/2 * Δᶠ * (𝒥ᴺᴺ * 𝒥ᴿᴺ)^(-1 / 8)
+    
+    L₁₁ = u̅u̅ - u̅ * u̅
+    L₂₂ = v̅v̅ - v̅ * v̅
+    L₃₃ = w̅w̅ - w̅ * w̅    
+    L₁₂ = u̅v̅ - u̅ * v̅
+    L₁₃ = u̅w̅ - u̅ * w̅
+    L₂₃ = v̅w̅ - v̅ * w̅
+
+    # Here we assume that α (ratio between scales) is 4 and
+    # β (ratio between the Smagorinsky coefficient at different scales)
+    # is one because the model is assumed to be scale - invariant
+    α = 4
+    β = 1
+
+    M₁₁ = 2Δᶠ * (S̅S̅₁₁ - α^2 * β^2 * S̅ₘ * S̅₁₁)
+    M₂₂ = 2Δᶠ * (S̅S̅₂₂ - α^2 * β^2 * S̅ₘ * S̅₂₂)
+    M₃₃ = 2Δᶠ * (S̅S̅₃₃ - α^2 * β^2 * S̅ₘ * S̅₃₃)
+    M₁₂ = 2Δᶠ * (S̅S̅₁₂ - α^2 * β^2 * S̅ₘ * S̅₁₂)
+    M₁₃ = 2Δᶠ * (S̅S̅₁₃ - α^2 * β^2 * S̅ₘ * S̅₁₃)
+    M₂₃ = 2Δᶠ * (S̅S̅₂₃ - α^2 * β^2 * S̅ₘ * S̅₂₃)
```

Comment:
Should this be

```suggestion
    M₁₁ = 2Δᶠ * (S̅S̅₁₁ - α^2 * β * S̅ₘ * S̅₁₁)
    M₂₂ = 2Δᶠ * (S̅S̅₂₂ - α^2 * β * S̅ₘ * S̅₂₂)
    M₃₃ = 2Δᶠ * (S̅S̅₃₃ - α^2 * β * S̅ₘ * S̅₃₃)
    M₁₂ = 2Δᶠ * (S̅S̅₁₂ - α^2 * β * S̅ₘ * S̅₁₂)
    M₁₃ = 2Δᶠ * (S̅S̅₁₃ - α^2 * β * S̅ₘ * S̅₁₃)
    M₂₃ = 2Δᶠ * (S̅S̅₂₃ - α^2 * β * S̅ₘ * S̅₂₃)
```
?

Reply:
Can you paste in the reference that shows this?

Reply:
Taking a second look at things I now think there may be more typos than this one. So I suggest @xkykai and @simone-silvestri double-check this, but if I understand correctly, what's called `𝒥ᴿᴺ` in this PR is called $\mathcal J^{QN}$in [Bou-Zeid et al. (2005)](https://dx.doi.org/10.1063/1.1839152), and here's its advection equation:

![image](https://github.com/CliMA/Oceananigans.jl/assets/13205162/c995259b-13c2-41e0-8f27-5689cc4c4ad0)

so using the language of this PR the forcing here should be
```julia
1 / T4Δ * (QᵢⱼNᵢⱼ - 𝒥ᴿᴺ)
```
instead of what's written below: `1 / TΔ * (LᵢⱼMᵢⱼ - 𝒥ᴿᴺ)`.

And the functions $Q_{ij}$ and $N_{ij}$ are:

![image](https://github.com/CliMA/Oceananigans.jl/assets/13205162/13848794-edb6-4f1a-8638-b41f2008e9df)

Similar comments are valid for the function below (`𝒥ᴺᴺ_forcing_function()`).

Reply:
For the record, my original comment had to do with $M_{ij}$ being implemented as `M₁₁ = 2Δᶠ * (S̅S̅₁₁ - α^2 * β^2 * S̅ₘ * S̅₁₁)` when it should be

![image](https://github.com/CliMA/Oceananigans.jl/assets/13205162/79887e48-e387-408b-b14d-f88dd828c8dc)

but now (per my previous comment) I realize that $M_{ij}$ isn't supposed to be in this forcing function at all.

Reply:
The difference between Q and L (and M and N) is the filter size. The stresses for the `𝒥ᴿᴺ` forcings are averaged twice as much as the `𝒥ᴸᴹ` forcings (if you look at how the averages are calculated). However, I would say that to push this PR forward we need to:
- define an averaging operators (to be held in the closure struct) for the `M - L` fluxes and for the `R - N` fluxes
- remove all this combination of interpolations in favor of those averaging operators

Reply:
> The difference between Q and L (and M and N) is the filter size.

Agreed. The $N$ and $M$ nomenclature difference makes it clearer though. That said, the combination $\alpha^2 \beta^2$ shouldn't appear anywhere, right? It's either $\alpha^4 \beta^2$ or $\alpha^2 \beta$ (unless you hard-coded some sutff I guess)

Reply:
What do the tilde, hat, and bar mean?

Reply:
- Tilde: LES-filtered velocity. In our case it's simply the velocities that come out of `model.velocities`
- Bar: First test filter
- Hat: Second test filter (at a larger spatial scale than the first)